core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1624,6 +1630,47 @@ private RexNode simplifyOrs(List<RexNode> terms, RexUnknownAs unknownAs) {
             continue;
           }
         }
+        break;
+      case NOT_EQUALS:
+        final Comparison notEqualsComparison =
+            Comparison.of(term, node -> RexUtil.isReferenceOrAccess(node, false));
+        if (notEqualsComparison != null) {
+          // We are dealing with a X<>A term, check if we saw before another NOT_EQUALS involving X
+          final RexNode prevNotEquals = notEqualsComparisonMap.get(notEqualsComparison.ref);
+          if (prevNotEquals == null) {
+            // This is the first NOT_EQUALS involving X, put it in the map
+            notEqualsComparisonMap.put(notEqualsComparison.ref, term);
+          } else {
+            // There is already in the map another NOT_EQUALS involving X:
+            //   - if it is already an IS_NOT_NULL: it was already simplified, ignore this term
+            //   - if it is not an IS_NOT_NULL (i.e. it is a NOT_EQUALS): check comparison values
+            if (prevNotEquals.getKind() != SqlKind.IS_NOT_NULL) {
+              final Comparable comparable1 = notEqualsComparison.literal.getValue();
+              //noinspection ConstantConditions
+              final Comparable comparable2 = Comparison.of(
+                  prevNotEquals, node -> RexUtil.isReferenceOrAccess(node, false))
+                  .literal.getValue();
+              //noinspection unchecked
+              if (comparable1.compareTo(comparable2) != 0) {
+                // We found: X <> A OR X <> B
+                if (unknownAs == RexUnknownAs.TRUE || !term.getType().isNullable()) {
```

Comment:
* you should choose one path or the other - but not both...if you go the generic way then let other parts of `RexSimplify` do the work; by replacing the term with the result of: `simplify( ref IS NOT NULL OR NULL )`
* ref IS NOT NULL will return false when the original expression should be null - please also add a testcase for this
```
    checkSimplify(
        or(
            ne(vInt(), literal(1)),
            ne(vInt(), literal(2))),
        "...");
```

Reply:
@kgyrtkirk , thanks for your feedback. 
I'm not sure I fully understand you here. What you mean is that I should replace:
`ref <> A OR ref <> B => ref IS NOT NULL OR NULL`
in all cases (without checking unknownAs value or ref's nullability), and let the other parts of RexSimplify to carry on with any potential further simplification?

Reply:
the expression: `ref IS NOT NULL OR NULL` is the logically equvivalent rewrite of `ref<> 1 or ref<>2`
instead of trying to guess in what conditions `ref IS NOT NULL OR NULL` will evaluate to true; run simplify on it; and if it gets back with an `isAlwaysTrue` ...then return true

for example:  even thru ref might be nullable - we may have a predicate which states that `ref is not null`; so it actually will be able to simplify it to true with other techniques

Reply:
Ok, thanks for the explanation. I have updated the PR trying to achieve what you describe.